packages/trader/src/AppV2/Components/TradeParameters/RiskManagement/take-profit-and-stop-loss-input.tsx

```diff
@@ -0,0 +1,328 @@
+import React from 'react';
+import clsx from 'clsx';
+import { observer } from 'mobx-react';
+import { useTraderStore } from 'Stores/useTraderStores';
+import { CONTRACT_TYPES, getCurrencyDisplayCode, getDecimalPlaces, useIsMounted, WS } from '@deriv/shared';
+import { focusAndOpenKeyboard } from 'AppV2/Utils/trade-params-utils';
+import { ActionSheet, CaptionText, Text, ToggleSwitch, TextFieldWithSteppers } from '@deriv-com/quill-ui';
+import { Localize, localize } from '@deriv/translations';
+import { requestPreviewProposal } from 'Stores/Modules/Trading/Helpers/preview-proposal';
+import { TTradeStore } from 'Types';
+import { getDisplayedContractTypes } from 'AppV2/Utils/trade-types-utils';
+import { ExpandedProposal } from 'Stores/Modules/Trading/Helpers/proposal';
+
+type TTakeProfitAndStopLossInputProps = {
+    classname?: string;
+    has_save_button?: boolean;
+    initial_error_text?: string;
+    onActionSheetClose: () => void;
+    parent_subscription_id_ref?: React.MutableRefObject<string | undefined>;
+    parent_ref?: React.MutableRefObject<{
+        has_take_profit?: boolean;
+        has_stop_loss?: boolean;
+        take_profit?: string;
+        tp_error_text?: string;
+        stop_loss?: string;
+        sl_error_text?: string;
+    }>;
+    parent_is_be_response_received_ref?: React.MutableRefObject<boolean>;
+    type?: 'take_profit' | 'stop_loss';
+    should_wrap_with_actionsheet?: boolean;
+};
+
+const TakeProfitAndStopLossInput = ({
+    classname,
+    has_save_button = true,
+    initial_error_text,
+    onActionSheetClose,
+    parent_subscription_id_ref,
+    parent_ref,
+    parent_is_be_response_received_ref,
+    type = 'take_profit',
+    should_wrap_with_actionsheet = true,
+}: TTakeProfitAndStopLossInputProps) => {
+    const trade_store = useTraderStore();
+    const {
+        contract_type,
+        currency,
+        has_take_profit,
+        has_stop_loss,
+        is_accumulator,
+        is_multiplier,
+        take_profit,
+        stop_loss,
+        trade_types,
+        trade_type_tab,
+        onChangeMultiple,
+        validation_params,
+    } = trade_store;
+    const isMounted = useIsMounted();
+
+    const is_take_profit_input = type === 'take_profit';
+
+    const subscription_id = React.useRef<string>();
+    const subscription_id_ref = parent_subscription_id_ref || subscription_id;
+
+    // For handling cases when user clicks on Save btn before we got response from BE
+    const is_be_response_received = React.useRef(false);
```

Comment:
Please use `api` and not `be` term because it's more clear

Reply:
And yeah it was mentioned above already, but this logic is very complicated. I thought it's a text field or something, but there are so many conditions, states and variables :-(

Reply:
Sure, will replace with api.
Regarding complicated logic: I 100% agree, but currently we have some limitations from BE side. You should understand, because you faced the same issue with getting available trade types from API.
I've already asked for some improvements and additional data in API response. After it will be done, I can refactor this part of code.